### PR TITLE
feat(junit-playwright)

### LINKED ISSUE #1369: [Feature] JUnit5 integration
To provide decent support of JUnit in codegen (as requested in https://github.com/microsoft/playwright-java/issues/1039), we need to first implement basic Playwright fixtures similar to other languages. This came up in the discussion of https://github.com/microsoft/playwright-java/pull/1366 and whether/how we can integrate https://github.com/uchagani/junit-playwright into Playwright. Below are some of the requirements we'll have for such integration and some comments regarding how it compares to the existing junit-playwright.

The execution model should be similar to [the default](https://playwright.dev/docs/test-parallel#parallelize-tests-in-a-single-file) execution model of Node.js Playwright. I.e. it should be possible to
 - configure browser options per test file, so that an instance of the browser is reused by all tests in the same file
 - do not parallel tests in the same file, only run in parallel tests in different classes (files), i.e. in terms of junit config:
                junit.jupiter.execution.parallel.enabled = true
                junit.jupiter.execution.parallel.mode.default = same_thread
                junit.jupiter.execution.parallel.mode.classes.default = concurrent

Builtin fixtures should create Playwright, Browser, BrowserContext and Page. Their lifetimes should match those in Node.js version, i.e.
  - new context and page are created before each test
  - for each test, same instance of page and context is used within the test, @beforeEach and @afterEach hooks 
  - page and context are closed right after last @afterEach hook finished for the test
 
Browser & playwright lifetimes should be bound to the thread & browser configuration. Playwright instance most likely should be thread local and can be reused by all tests running on that thread. We should try to reuse the same browser instance for running more tests that require the browser launch options. Currently, playwright internal test suite creates new browser per test class which seems to be good enough.

To keep things simple in the beginning we'd like to replace PlaywrightBrowserConfig with a configurable BrowserFactory, so that we don't have to think about config format and let the user create custom browser themselves. Eventually, we'll likely want to load the config from a json/xml/.properties file. We'll likely need to experiment with a few approaches here, to strike a balance.

### PR DESCRIPTION
**Pull request by uchagani**
Resolves #1369 

@yury-s here is an initial implementation of `Playwright` and `Browser` fixtures.  This implementation works for both `concurrent` and `same_thread` modes.  I will add `BrowserContext` and `Page` implementations after your feedback.

Notes:
- I added comments where needed to explain things but will be removed when reviewed.
- I made changes to the `Browser` API to do some testing.  While I think these would be good to have, they will be removed once reviewed.
- The tests can probably be removed if we decide to use this in the actual playwright-java tests but for now  i needed to see if everything is working so i added these tests

### CODE REVIEW

File: playwright-junit/src/main/java/com/microsoft/playwright/BrowserExtension.java
Comment (yury-s) [inline playwright-junit/src/main/java/com/microsoft/playwright/BrowserExtension.java:73]: We might want to cache `BrowserFactory` instance and create it just once per thread/run etc. For now creating all it every time is fine.

File: playwright-junit/src/main/java/com/microsoft/playwright/BrowserExtension.java
Comment (yury-s) [inline playwright-junit/src/main/java/com/microsoft/playwright/BrowserExtension.java:51]: Let's close the browser in afterAll callback in this class. Also if you wan to use per method annotation for browser factory, the browser should be closed after such test method finished running (aka afterEach). But I'd keep things simple for now and only support class-level `UseBrowserFactory` annotations, WDYT?

File: playwright-junit/src/main/java/com/microsoft/playwright/PlaywrightExtension.java
Comment (yury-s) [inline playwright-junit/src/main/java/com/microsoft/playwright/PlaywrightExtension.java:43]: Ideally we'd keep `Playwright` instances until all the tests finished running. Something like [RunListener.testRunFinished](https://junit.org/junit4/javadoc/4.12/org/junit/runner/notification/RunListener.html#testRunFinished(org.junit.runner.Result)) seems a good hook to run such logic but it looks like that interface was removed in JUnit 5. Maybe there is a new counterpart?

File: playwright-junit/src/main/java/com/microsoft/playwright/ExtensionUtils.java
Comment (uchagani) [reply playwright-junit/src/main/java/com/microsoft/playwright/ExtensionUtils.java:15]: We would still need this method because of the bug I referenced unfortunately. 

Supporting per method is not adding any extra complexity aside from knowing when to close the objects. 

The creation of all objects, regardless of happens when the ParameterResolver runs and that would be up to the user whether they need it in a hook or the test method.

File: playwright-junit/src/main/java/com/microsoft/playwright/PlaywrightExtension.java
Comment (uchagani) [reply playwright-junit/src/main/java/com/microsoft/playwright/PlaywrightExtension.java:31]: I made these public so users could build their own extensions on top of junit-playwright.

For example, creating auto-navigation extensions or auto-login extensions for their test frameworks.

For now, we can make it private if you'd like but I think this is something that we should allow at some point.

File: playwright-junit/src/main/java/com/microsoft/playwright/PlaywrightCleanupExtension.java
Comment (uchagani) [reply playwright-junit/src/main/java/com/microsoft/playwright/PlaywrightCleanupExtension.java:12]: We can but then the registration of extensions comes into play regarding which AfterAllCallback hook runs first (we would want browser then playwright in that order). Which is okay since we are registering extensions with our custom annotation.

File: playwright-junit/src/main/java/com/microsoft/playwright/BrowserExtension.java
Comment (uchagani) [reply playwright-junit/src/main/java/com/microsoft/playwright/BrowserExtension.java:73]: Makes sense, should be doable with another map.

File: playwright-junit/src/main/java/com/microsoft/playwright/BrowserExtension.java
Comment (uchagani) [reply playwright-junit/src/main/java/com/microsoft/playwright/BrowserExtension.java:51]: I think we don't don't currently want to support method-level parallelizations then we shouldn't support method-level custom factory. That way we don't have to deal with any after each callbacks

File: playwright-junit/src/main/java/com/microsoft/playwright/PlaywrightExtension.java
Comment (uchagani) [reply playwright-junit/src/main/java/com/microsoft/playwright/PlaywrightExtension.java:43]: Junit 5 doesn't have a built in after test run hook.

They have a workaround using a closeable resource but it currently could run in a separate thread than the class which means we couldn't access the threadlocal objects.

I've opened an issue with JUnit for this also.

For now, I think after class is our best option.

**Comment by uchagani** [comment]
@yury-s removed all method-level parallelization code and made the other updates you suggested.  If you like this approach, I can continue with BrowserContext and Page fixtures.

File: playwright-junit/src/main/java/com/microsoft/playwright/junit/BrowserExtension.java
Comment (yury-s) [inline playwright-junit/src/main/java/com/microsoft/playwright/junit/BrowserExtension.java:71]: How can it throw if `supportsParameter` is called first and checks that the annotation is present?

File: playwright-junit/src/test/java/com/microsoft/playwright/junit/BrowserDefaultFactoryTests.java
Comment (yury-s) [inline playwright-junit/src/test/java/com/microsoft/playwright/junit/BrowserDefaultFactoryTests.java:24]: Is it guaranteed that @afterAll of BrowserExtension is called after all afterAll's in the test class?

File: playwright-junit/src/main/java/com/microsoft/playwright/junit/BrowserExtension.java
Comment (yury-s) [inline playwright-junit/src/main/java/com/microsoft/playwright/junit/BrowserExtension.java:16]: Can there be more than one entry in the map? My understanding is that there is at most one active test class and each class has only one UseBrowserFactory annotation.

File: playwright-junit/src/test/java/com/microsoft/playwright/junit/BrowserCustomFactoryTests.java
Comment (yury-s) [inline playwright-junit/src/test/java/com/microsoft/playwright/junit/BrowserCustomFactoryTests.java:8]: I wonder if it'd be more appropriate to make this a parameter/field annotation. It'd be probably more verbose but also more explicit, what are you thoughts?

**Review by yury-s — COMMENTED** [review]
Left a few comments, but I like this much better. I think we should continue in this direction.

File: playwright-junit/src/main/java/com/microsoft/playwright/junit/BrowserExtension.java
Comment (uchagani) [reply playwright-junit/src/main/java/com/microsoft/playwright/junit/BrowserExtension.java:71]: It can't, I just thought but would be better than null but I'll look at this again to find a better way.

File: playwright-junit/src/main/java/com/microsoft/playwright/junit/BrowserExtension.java
Comment (uchagani) [reply playwright-junit/src/main/java/com/microsoft/playwright/junit/BrowserExtension.java:16]: Good catch. This is a left over from when we would allow more than one browser per class. I'll fix it.

File: playwright-junit/pom.xml
Comment (jfgreffier) [inline playwright-junit/pom.xml:12]: It would be nice to also have a <name> and <description>

File: playwright-junit/src/test/java/com/microsoft/playwright/junit/BrowserCustomFactoryTests.java
Comment (uchagani) [reply playwright-junit/src/test/java/com/microsoft/playwright/junit/BrowserCustomFactoryTests.java:8]: Are you thinking something like this:

```java
@Test
void foo(@UseBrowserFactory(CustomBrowser.class) Page page) {

  //test

}
```

Personally I think class level annotation is good as it is saying the at all tests in this class will use this browser factory.

We can allow further customizing in future PRs.

Comment (yury-s) [reply playwright-junit/src/test/java/com/microsoft/playwright/junit/BrowserCustomFactoryTests.java:8]: Yes, I meant something like @TempDir junit annotation. I agree that having one class-level annotation is more convenient, let's stick with that, can definitely be customized in the future.

File: playwright-junit/src/main/java/com/microsoft/playwright/junit/PlaywrightExtension.java
Comment (yury-s) [inline playwright-junit/src/main/java/com/microsoft/playwright/junit/PlaywrightExtension.java:32]: Remove logging here and in other places?

File: playwright-junit/src/test/java/com/microsoft/playwright/junit/BrowserCustomFactoryTests.java
Comment (yury-s) [inline playwright-junit/src/test/java/com/microsoft/playwright/junit/BrowserCustomFactoryTests.java:24]: Remove debug logging.

File: playwright-junit/src/test/java/com/microsoft/playwright/junit/BrowserCustomFactoryTests.java
Comment (yury-s) [inline playwright-junit/src/test/java/com/microsoft/playwright/junit/BrowserCustomFactoryTests.java:30]: `assertNotNull(browser);` here and in other places so that we use junit assertions in the tests and not rely on java assertions to be enabled?

File: playwright-junit/src/test/java/com/microsoft/playwright/junit/BrowserDefaultFactoryTests.java
Comment (yury-s) [inline playwright-junit/src/test/java/com/microsoft/playwright/junit/BrowserDefaultFactoryTests.java:9]: As we rely on ThreadLocal, can we also add a test for the fixture running on 2 threads concurrently?

File: playwright-junit/src/test/java/com/microsoft/playwright/junit/PlaywrightTests.java
Comment (yury-s) [inline playwright-junit/src/test/java/com/microsoft/playwright/junit/PlaywrightTests.java:53]: What if someone has (intentionally or by accident) 2 Playwright parameters, will it be one and the same instance? Let's add a test for that. Same for the Browser fixture.

**Review by yury-s — COMMENTED** [review]
> I will add BrowserContext and Page implementations after your feedback.

Do you want to add them to the same PR or follow-up in a separate one?

File: playwright-junit/src/test/java/com/microsoft/playwright/junit/BrowserCustomFactoryTests.java
Comment (yury-s) [reply playwright-junit/src/test/java/com/microsoft/playwright/junit/BrowserCustomFactoryTests.java:8]: One thing that is not very clear with class level annotation is fields with type `Browser`, but it's probably fine if they are not assigned by the fixture.

File: playwright-junit/src/test/java/com/microsoft/playwright/junit/BrowserCustomFactoryTests.java
Comment (uchagani) [reply playwright-junit/src/test/java/com/microsoft/playwright/junit/BrowserCustomFactoryTests.java:30]: I was thinking of removing these tests altogether and using playwright-junit in the actual playwright tests.  what do you think?  that way we won't need a separate test suite for this.  If we do want a separate test suite, JUnit provides a way to test extensions that we could use to properly test the extensions.

Comment (yury-s) [reply playwright-junit/src/test/java/com/microsoft/playwright/junit/BrowserCustomFactoryTests.java:30]: That'd be nice and would also help us evaluate usability of the fixtures.

Comment (uchagani) [reply playwright-junit/src/test/java/com/microsoft/playwright/junit/BrowserCustomFactoryTests.java:30]: @yury-s I'm running into a circular dependency issue when trying to update the playwright tests to using playwright-junit.

Since playwright-junit needs playwright as a dependency, do you know of a clean way to use playwright-junit in playwright?  

i can think of 2 ways:

1.  include junit-playwright inside of the main playwright package but with the junit dependency as provided scope.  
2. put junit-playwright in a completely separate maven project

What are your thoughts?

Comment (yury-s) [reply playwright-junit/src/test/java/com/microsoft/playwright/junit/BrowserCustomFactoryTests.java:30]: If I understand correctly, marking `playwright` dependency of `playwright-junit` `provided` should enable using it in the `playwright` tests, is it not working? I believe you need to remove `playwright` from the parent pom.xml in that case.

Comment (yury-s) [reply playwright-junit/src/test/java/com/microsoft/playwright/junit/BrowserCustomFactoryTests.java:30]: If this does not work, including junit stuff into the main playwright module with "provided" junit dependency is okay too.

Comment (uchagani) [reply playwright-junit/src/test/java/com/microsoft/playwright/junit/BrowserCustomFactoryTests.java:30]: Yea i don't think it will work.  If i remove playwright from the parent pom, then maven complains it can't find it inside `playwright-junit`.  And once i add `playwright-junit` to `playwright`, i get

```
java: Annotation processing is not supported for module cycles. Please ensure that all modules from cycle [playwright,playwright-junit] are excluded from annotation processing
```

Comment (uchagani) [reply playwright-junit/src/test/java/com/microsoft/playwright/junit/BrowserCustomFactoryTests.java:30]: I'll just include it in the main playwright module

**Comment by uchagani** [comment]
@yury-s this is ready for another review.  All playwright fixtures have been added.  I also started updating the tests to use the fixtures.  Please let me know if updating the tests in the manner that I did it is okay, and if so, i can continue updating the rest of them.  

The idea behind the tests is to use the `TestBase` only for the server parts of the tests.  all other playwright-related things should be handled by the fixtures.  See my comments in the test classes themselves

File: playwright/src/test/java/com/microsoft/playwright/TestBrowserContextBasic.java
Comment (uchagani) [inline playwright/src/test/java/com/microsoft/playwright/TestBrowserContextBasic.java:38]: @yury-s this and another test below are failing for the same reason.  If using fixtures, i think these assertions should be updated to be 0 for the context list size.

File: playwright/src/test/java/com/microsoft/playwright/TestBrowserContextBasic.java
Comment (uchagani) [inline playwright/src/test/java/com/microsoft/playwright/TestBrowserContextBasic.java:101]: here is the other place it is failing because of the old TestBase's default browser context creation.

Comment (yury-s) [reply playwright/src/test/java/com/microsoft/playwright/TestBrowserContextBasic.java:101]: Yes, you are right. Previously we were creating a context unconditionally in beforeEach and we assert that the closed contexts are removed from the contexts() list.

File: playwright/src/test/java/com/microsoft/playwright/TestBrowserContextBasic.java
Comment (yury-s) [reply playwright/src/test/java/com/microsoft/playwright/TestBrowserContextBasic.java:38]: Yes, updating to 0 sounds right, see my response below.

**Review by yury-s — COMMENTED** [review]
> Please let me know if updating the tests in the manner that I did it is okay, and if so, i can continue updating the rest of them.

Updated tests look good, but since it is going to be a lot of mostly mechanical changes in many files I suggest we do it in a separate PR and keep this one focused on the implementation. Having just a few tests that you had in the previous iteration should be fine for the initial implementation and should unblock this PR.

File: playwright/src/main/java/com/microsoft/playwright/junit/BrowserExtension.java
Comment (uchagani) [reply playwright/src/main/java/com/microsoft/playwright/junit/BrowserExtension.java:11]: If we move the files into another package we will need to make the extensions public instead of package-private.  Do we want to expose them?

File: playwright/src/main/java/com/microsoft/playwright/junit/DefaultPlaywrightFactory.java
Comment (uchagani) [reply playwright/src/main/java/com/microsoft/playwright/junit/DefaultPlaywrightFactory.java:6]: They don't have to be public since they are only used internally.  but we can make them public if you'd like

File: playwright/src/test/java/com/microsoft/playwright/BrowserFromEnv.java
Comment (uchagani) [reply playwright/src/test/java/com/microsoft/playwright/BrowserFromEnv.java]: Do we want to use `HEADLESS` instead so that the API to configure tests and for playwright are the same?

**Comment by uchagani** [comment]
@yury-s left some more comments, for some reason my comments before were showing up as pending and i had to submit the review.

File: playwright/src/test/java/com/microsoft/playwright/BrowserFromEnv.java
Comment (yury-s) [reply playwright/src/test/java/com/microsoft/playwright/BrowserFromEnv.java]: `HEADED` is preferred as it defaults to `false` when not specified (rather than headless that is true by default and only if specified as HEADLESS=0 it actually changes the behavior), consistent with --headed in node.js command line and  `HEADED` in .Net. We didn't change the parameter in the library API as it had already been shipped by the time we realized this in the test runner.

File: playwright/src/main/java/com/microsoft/playwright/junit/DefaultPlaywrightFactory.java
Comment (yury-s) [reply playwright/src/main/java/com/microsoft/playwright/junit/DefaultPlaywrightFactory.java:6]: > They don't have to be public since they are only used internally. 

Can people use use the explicitly in the annotations today (`@UsePlaywright(playwright = DefaultPlaywrightFactory.class)`) ?

> but we can make them public if you'd like

I see 2 reasons to make them public: people may want to subclass them, they may want to use them explicitly in the annotations.

File: playwright/src/main/java/com/microsoft/playwright/junit/BrowserExtension.java
Comment (yury-s) [reply playwright/src/main/java/com/microsoft/playwright/junit/BrowserExtension.java:11]: Yeah, I think it is fine, although not ideal. There are already precedents of this, e.g. PlaywrightImpl. We don't make any promises about backwards compatibility of the classes in the `impl` package, those who use them are on their own.

**Comment by yury-s** [comment]
> @yury-s left some more comments, for some reason my comments before were showing up as pending and i had to submit the review.

Yeah, that's sometimes annoying that they are not published immediately.

File: playwright/src/main/java/com/microsoft/playwright/junit/DefaultPlaywrightFactory.java
Comment (uchagani) [reply playwright/src/main/java/com/microsoft/playwright/junit/DefaultPlaywrightFactory.java:6]: Sure I can make them public. I figured no one would explicitly use the default ones and would define their own if they wanted something custom.

**Comment by uchagani** [comment]
@yury-s addressed all previous suggestions.  ready for another review.

File: playwright/src/main/java/com/microsoft/playwright/junit/impl/BrowserExtension.java
Comment (yury-s) [inline playwright/src/main/java/com/microsoft/playwright/junit/impl/BrowserExtension.java:19]: ditto

**Review by yury-s — APPROVED** [review]

File: playwright/src/main/java/com/microsoft/playwright/junit/impl/BrowserContextExtension.java
Comment (uchagani) [reply playwright/src/main/java/com/microsoft/playwright/junit/impl/BrowserContextExtension.java:83]: If we drop it we would have to recreate the factory for each test. The way we currently have it, we only need to create the factory the first time and reuse it for the entire test class.

Comment (yury-s) [reply playwright/src/main/java/com/microsoft/playwright/junit/impl/BrowserContextExtension.java:83]: I thought it was already happening when `cleanupBrowserContextFactory` is called from `afterEach`, basically the thread local cache is cleared between the test runs. Am I missing something?

Comment (uchagani) [reply playwright/src/main/java/com/microsoft/playwright/junit/impl/BrowserContextExtension.java:83]: That is a good catch, it should not be cleaned up in the `afterEach` but in an `afterAll`

File: playwright/src/test/java/com/microsoft/playwright/TestFixturesWithCustomFactories.java
Comment (uchagani) [reply playwright/src/test/java/com/microsoft/playwright/TestFixturesWithCustomFactories.java:30]: i'm not sure how we can do this without using JUnit's extension testing engine.

Comment (yury-s) [reply playwright/src/test/java/com/microsoft/playwright/TestFixturesWithCustomFactories.java:30]: Ok, never mind then. I thought about something simple like a static/thread-local counter field inside CustomBrowserFactory and a check in afterAll of the test. We can certainly live without it.

**Comment by uchagani** [comment]
@yury-s made the latest changes:  

* `BrowserContextFactory` is now cleaned up in `AfterAllCallback` instead of `AfterEachCallback`
* invalid `BROWSER` value will now throw.

File: playwright/src/main/java/com/microsoft/playwright/junit/impl/BrowserContextExtension.java
Comment (yury-s) [inline playwright/src/main/java/com/microsoft/playwright/junit/impl/BrowserContextExtension.java:87]: nit: `threadLocalBrowserContextFactory.remove();` can be called unconditionally as it is noop when there is no value.

File: playwright/src/main/java/com/microsoft/playwright/junit/impl/BrowserContextExtension.java
Comment (yury-s) [inline playwright/src/main/java/com/microsoft/playwright/junit/impl/BrowserContextExtension.java:76]: nit: maybe restructure it like this to avoid try/finally?

```java
      BrowserContext browserContext = threadLocalBrowserContext.get();
      threadLocalBrowserContext.remove();
      if (browserContext != null) {
        browserContext.close();
      }
```

File: playwright/src/main/java/com/microsoft/playwright/junit/impl/BrowserExtension.java
Comment (yury-s) [inline playwright/src/main/java/com/microsoft/playwright/junit/impl/BrowserExtension.java:64]: nit: this method is called just once per class when Browser instance is being created (and cached), there is not much point in caching the factory I believe.

**Review by yury-s — APPROVED** [review]
A few nits, but I'm okay with merging in its current form.

**Comment by uchagani** [comment]
@yury-s can I get another approval. Had to fix a test.

**Review by yury-s — APPROVED** [review]

**Comment by yury-s** [comment]
`TestFixturesWithDefaultFactories.beforeAll:20 expected: <null> but was: <chromium>` for some reason is failing in Docker, the other failures on the bots look unrelated.

**Comment by uchagani** [comment]
Could be in docker we aren't setting `BROWSER`. I can update the test to account for that.

**Comment by yury-s** [comment]
> Could be in docker we aren't setting `BROWSER`. I can update the test to account for that.

We don't.

**Comment by uchagani** [comment]
@yury-s, fixed.

**Review by yury-s — APPROVED** [review]

**Pull request merged**
